docs/389ds/design/worker-threads.md

```diff
@@ -0,0 +1,345 @@
+
+--------------------------------------------------------------------
+
+{% include toc.md %}
+
+Introduction
+------------
+
+The current worker threads (that process the ldap operations) implementation has several flaws:
+
+- need to restart the server after changing the configured number of threads
+- contention at the global mutex and condition variable level
+- configuring too few or too many worker threads noticeably impact the performance.
+
+To fix these flaws, the new implementation objectives are:
+
+- Allow to change the number of worker threads without needing to restart the server.  (keeping the current configuration attribute and cli command)
+- Fix the condition variable contention issue.
+- Reduce the impact of having too many worker threads configured
+- Provide some monitoring data.
+
+This proposal achieves that by:
+
+- Allowing to create and stop worker thread dynamically.
+- Not associating a condition variable with the global mutex but have a specific mutex and condition variable for each worker threads.
+- Waking up a worker thread only if it has something to do.
+- Providing some monitoring data.
+
+## Current implementation synopsys
+
+--------------------
+
+There is a global mutex and a global condition variable and a global queue
+
+- listener threads push data in in the global queue and wake up the worker threads
+- worker threads loops waiting on the global condition variable for new work
+  then process it.
+- to speed up the queueing and limit the alloc/free a free list element is used
+
+## New implementation
+
+--------------------
+
+### Stacked data
+
+In current implementation there are two stacks to handle:
+
+- the work queue element
+- the operation struct
+  and avoid freeing/allocing them every time.
+
+I instead proposed to:
+
+- keep the operation in the thread context. (Anyway a thread only works on a single operation at a time)
+  - keep a free list of element for the waiting_job list. (That cannot be more than the maximum of connections in the queue so in the worse case the free list will 3*8*64K (i.e around 3 Mb) but I wonder if we should not reduce this list size and let the listening threads sleep for a few ms if there are no more items. ( Anyway if the sleep delay is smaller than the time needed to flush the waiting job queue, no time is wasted.)
+
+### Overload warning (flow control)
+
+A warning is logged if
+
+- The warning has not been logged since some delay (1 minute)
+- The waiting job queue size gets higher than a threshold (100*number of worker threads)
+
+### Producters/consumers algorithm
+
+There is still a global mutex and also three lists:
+
+- waiting_threads
+- busy_threads
+- waiting_jobs
+  Access to these lists are protected by the global mutex lock
+
+When pushing new job the listener threads:
+
+- Get global mutex
+
+- Loop forever:
+  
+  - if waiting_threads is not empty:
+    
+    - Select the first waiting thread
```

Comment:
I wonder if we could get rid of global lock to manage waiting_thread and busy_thread. For example if we allocate two fixed size (e.g. 1000) arrays of waiting_thread/busy_thread. A worker is either in slot 'i' of waiting_thread or in slot 'i' busy_thread. Then iterate over waiting_thread with atomic set/get. If a listener atomic_get/set a thread in waiting_threads, it books it and move it in atomic_get/set busy_thread.

The drawback is that we have a fixed maximum worker (e.g. 1000) but it could be configurable. The benefit would be less contention on the global_mutex.

A status (atomic_get/set) of a thread in the arrays is free/busy/not_allocated. In such case a single array would work as well.

For example, waiting/busy arrays have 1000 slots, 50 workers are allocated in the first 50 slots (out of 1000) of waiting/busy. When the number of workers increase with 10 new workers the allocated workers are stored in slots 51-60

Reply:
Not sure that iterating on an array using atomic operation for each elements is faster than a global mutex 
   (I am pretty sure it is worse).
I do not expect too much contention on that lock anyway: 
  - The number of listener thread is very limited
  - The worker threads needs only the lock when taking/finishing a job which is quite limited compared to the job duration (so even with 1000 threads contention should not waste too much time)
  - The problem with the original mutex was the nspr condition variable that wakes up all the threads at the same time and generating useless contention.
  And finally my feeling is that we have better see what are result with the current design and improve things later on if we thing

Reply:
I agree, several atomics are for sure much expensive that acquiring a single lock and on the contention side global lock should not be hot (to confirm with extrem load).